File: src/shared/kartEngine/KartShared/UnityEngine/Physics.ts
/**
 * UnityEngine.Physics类的TypeScript实现
 * 模拟Unity的Physics静态类功能
 */

import { Workspace, CollectionService } from "@rbxts/services";
import { LayerMask } from "./LayerMask";
import { RaycastHit } from "./RaycastHit";

export class Physics {
    // 模拟Unity的Physics.RaycastAll方法
    static RaycastAll(origin: Vector3, direction: Vector3, maxDistance?: number, layerMask?: number): RaycastHit[] {
        maxDistance = maxDistance || math.huge;
        layerMask = layerMask || 0xFFFFFFFF;

        const raycastParams = new RaycastParams();
        raycastParams.FilterType = Enum.RaycastFilterType.Blacklist;

        // 排除Player标签的物体（玩家赛车）
        const filterList: Instance[] = [];
        const playerObjects = CollectionService.GetTagged("Player");
        for (const obj of playerObjects) {
            filterList.push(obj);
        }
        raycastParams.FilterDescendantsInstances = filterList;

        // 转换为Roblox Vector3
        const robloxOrigin = new Vector3(origin.X, origin.Y, origin.Z);
        // 注意：direction可能已经是完整的向量，需要根据其长度判断
        const dirLength = math.sqrt(direction.X * direction.X + direction.Y * direction.Y + direction.Z * direction.Z);
        let robloxDirection: Vector3;
        if (dirLength > 1.5) {
            // direction已经是完整的向量（从GetCollisionCheckedPosition传入的是toPosition - fromPosition）
            robloxDirection = new Vector3(direction.X, direction.Y, direction.Z);
        } else {
            // direction是单位向量，需要乘以maxDistance
            robloxDirection = new Vector3(direction.X, direction.Y, direction.Z).mul(maxDistance);
        }

        const results: RaycastHit[] = [];
        const hitObjects: Instance[] = [];  // 已经击中的物体

        // 执行多次射线检测以模拟RaycastAll
        for (let i = 0; i < 50; i++) {
            // 合并过滤列表：Player标签物体 + 已击中的物体
            const currentFilterList: Instance[] = [];
            for (const obj of filterList) {
                currentFilterList.push(obj);
            }
            for (const obj of hitObjects) {
                currentFilterList.push(obj);
            }
            raycastParams.FilterDescendantsInstances = currentFilterList;

            const raycastResult = Workspace.Raycast(robloxOrigin, robloxDirection, raycastParams);

            if (raycastResult) {
                const hit = new RaycastHit(raycastResult);

                // 检查layer mask
                if (LayerMask.CheckLayerMask(hit.layer, layerMask)) {
                    results.push(hit);
                }

                hitObjects.push(raycastResult.Instance);
            } else {
                break;
            }
        }

        return results;
    }

    // 重载版本：接受Ray参数
    static RaycastAllFromRay(ray: { origin: Vector3; direction: Vector3 }, maxDistance?: number, layerMask?: number): RaycastHit[] {
        return this.RaycastAll(ray.origin, ray.direction, maxDistance, layerMask);
    }

    // 模拟Unity的Physics.Raycast方法（单次射线检测）
    static Raycast(origin: Vector3, direction: Vector3, maxDistance?: number, layerMask?: number): [boolean, RaycastHit?] {
        const results = this.RaycastAll(origin, direction, maxDistance, layerMask);
        return results.size() > 0 ? [true, results[0]] : [false, undefined];
    }
}

export default Physics;

File: sources/roblox-lua-items-network/shared/kartEngine/KartShared/UnityEngine/Physics.lua
-- UnityEngine.Physics类的Roblox实现
-- 模拟Unity的Physics静态类功能

local Physics = {}

local LayerMask = require(script.Parent.LayerMask)
local RaycastHit = require(script.Parent.RaycastHit)
local Ray = require(script.Parent.Ray)

-- Roblox服务
local Workspace = game:GetService("Workspace")
local CollectionService = game:GetService("CollectionService")

-- 模拟Unity的Physics.RaycastAll方法
function Physics.RaycastAll(origin, direction, maxDistance, layerMask)
    maxDistance = maxDistance or math.huge
    layerMask = layerMask or 0xFFFFFFFF
    
    local raycastParams = RaycastParams.new()
    raycastParams.FilterType = Enum.RaycastFilterType.Blacklist
    
    -- 排除Player标签的物体（玩家赛车）
    local filterList = {}
    local playerObjects = CollectionService:GetTagged("Player")
    for _, obj in ipairs(playerObjects) do
        table.insert(filterList, obj)
    end
    raycastParams.FilterDescendantsInstances = filterList
    
    -- 转换为Roblox Vector3
    local robloxOrigin = Vector3.new(origin.X, origin.Y, origin.Z)
    -- 注意：direction可能已经是完整的向量，需要根据其长度判断
    local dirLength = math.sqrt(direction.X * direction.X + direction.Y * direction.Y + direction.Z * direction.Z)
    local robloxDirection
    if dirLength > 1.5 then
        -- direction已经是完整的向量（从GetCollisionCheckedPosition传入的是toPosition - fromPosition）
        robloxDirection = Vector3.new(direction.X, direction.Y, direction.Z)
    else
        -- direction是单位向量，需要乘以maxDistance
        robloxDirection = Vector3.new(direction.X, direction.Y, direction.Z) * maxDistance
    end
    
    local results = {}
    local hitObjects = {}  -- 已经击中的物体
    
    -- 执行多次射线检测以模拟RaycastAll
    for i = 1, 50 do
        -- 合并过滤列表：Player标签物体 + 已击中的物体
        local currentFilterList = {}
        for _, obj in ipairs(filterList) do
            table.insert(currentFilterList, obj)
        end
        for _, obj in ipairs(hitObjects) do
            table.insert(currentFilterList, obj)
        end
        raycastParams.FilterDescendantsInstances = currentFilterList
        
        local raycastResult = Workspace:Raycast(robloxOrigin, robloxDirection, raycastParams)
        
        
        if raycastResult then
            local hit = RaycastHit.new(raycastResult)
            
            -- 检查layer mask
            if LayerMask.CheckLayerMask(hit.layer, layerMask) then
                table.insert(results, hit)
            end
            
            table.insert(hitObjects, raycastResult.Instance)
        else
            break
        end
    end
    
    return results
end

-- 重载版本：接受Ray参数
function Physics.RaycastAllFromRay(ray, maxDistance, layerMask)
    return Physics.RaycastAll(ray.origin, ray.direction, maxDistance, layerMask)
end

-- 模拟Unity的Physics.Raycast方法（单次射线检测）
function Physics.Raycast(origin, direction, maxDistance, layerMask)
    local results = Physics.RaycastAll(origin, direction, maxDistance, layerMask)
    return #results > 0, results[1]
end

return Physics